cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp

```diff
@@ -78,11 +84,108 @@ class HDFSFileReadBufferBuilder : public ReadBufferBuilder
         if (file_uri.getPort())
             uri_path += ":" + std::to_string(file_uri.getPort());
         DB::ReadSettings read_settings;
-        read_buffer = std::make_unique<DB::ReadBufferFromHDFS>(
-            uri_path, file_uri.getPath(), context->getGlobalContext()->getConfigRef(),
-            read_settings);
+        if (set_read_util_position) 
+        {
+            std::pair<size_t, size_t> start_end_pos = adjustFileReadStartAndEndPos(file_info.start(), file_info.start() + file_info.length(),
+                    uri_path, file_uri.getPath());
+            LOG_DEBUG(&Poco::Logger::get("ReadBufferBuilder"), "File read start and end position adjusted from {},{} to {},{}",
+                    file_info.start(), file_info.start() + file_info.length(), start_end_pos.first, start_end_pos.second);
+            read_buffer = std::make_unique<DB::ReadBufferFromHDFS>(
+                uri_path, file_uri.getPath(), context->getGlobalContext()->getConfigRef(),
+                read_settings, start_end_pos.second);
+            if (auto * seekable_in = dynamic_cast<DB::SeekableReadBuffer *>(read_buffer.get()))
+            {
+                seekable_in->seek(start_end_pos.first, SEEK_SET);
+            }
+        }
+        else
+        {
+            read_buffer = std::make_unique<DB::ReadBufferFromHDFS>(
+                uri_path, file_uri.getPath(), context->getGlobalContext()->getConfigRef(),
+                read_settings);
+        }
         return read_buffer;
     }
+
+    std::pair<size_t, size_t> adjustFileReadStartAndEndPos(
```

Comment:
Should use `ReadBuffer` instead of hdfs api direclty ?

Reply:
ReadBuffer don't have apis to adjust file position, and this should be better in HDFS ReadBuffer. @lgbo-ustc